File: src/com/android/systemui/plugin/globalactions/wallet/WalletPanelViewController.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.plugin.globalactions.wallet;

import android.app.PendingIntent;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.Icon;
import android.os.Handler;
import android.os.Looper;
import android.service.quickaccesswallet.GetWalletCardsError;
import android.service.quickaccesswallet.GetWalletCardsRequest;
import android.service.quickaccesswallet.GetWalletCardsResponse;
import android.service.quickaccesswallet.QuickAccessWalletClient;
import android.service.quickaccesswallet.SelectWalletCardRequest;
import android.service.quickaccesswallet.WalletCard;
import android.service.quickaccesswallet.WalletServiceEvent;
import android.util.Log;
import android.view.View;
import android.widget.FrameLayout;

import com.android.systemui.plugins.GlobalActionsPanelPlugin;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class WalletPanelViewController implements
        GlobalActionsPanelPlugin.PanelViewController,
        WalletCardCarousel.OnSelectionListener,
        QuickAccessWalletClient.OnWalletCardsRetrievedCallback,
        QuickAccessWalletClient.WalletServiceEventListener {

    private static final String TAG = "WalletPanelViewCtrl";
    private static final int MAX_CARDS = 10;
    private static final long SELECTION_DELAY_MILLIS = TimeUnit.SECONDS.toMillis(30);

    private final Context mPluginContext;
    private final QuickAccessWalletClient mWalletClient;
    private final WalletView mWalletView;
    private final WalletCardCarousel mWalletCardCarousel;
    private final GlobalActionsPanelPlugin.Callbacks mPluginCallbacks;
    private final ExecutorService mExecutor;
    private final Handler mHandler;
    private final Runnable mSelectionRunnable = this::selectCard;
    private boolean mIsDeviceLocked;
    private boolean mIsDismissed;
    private boolean mHasRegisteredListener;
    private String mSelectedCardId;

    public WalletPanelViewController(
            Context pluginContext,
            QuickAccessWalletClient walletClient,
            GlobalActionsPanelPlugin.Callbacks pluginCallbacks,
            boolean isDeviceLocked) {
        mPluginContext = pluginContext;
        mWalletClient = walletClient;
        mPluginCallbacks = pluginCallbacks;
        mIsDeviceLocked = isDeviceLocked;
        mWalletView = new WalletView(pluginContext);
        mWalletView.setLayoutParams(
                new FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        FrameLayout.LayoutParams.WRAP_CONTENT));
        mWalletCardCarousel = mWalletView.getCardCarousel();
        mWalletCardCarousel.setSelectionListener(this);
        mHandler = new Handler(Looper.myLooper());
        mExecutor = Executors.newSingleThreadExecutor();
    }

    /**
     * Implements {@link GlobalActionsPanelPlugin.PanelViewController}. Returns the {@link View}
     * containing the Quick Access Wallet.
     */
    @Override
    public View getPanelContent() {
        return mWalletView;
    }

    /**
     * Implements {@link GlobalActionsPanelPlugin.PanelViewController}. Invoked when the view
     * containing the Quick Access Wallet is dismissed.
     */
    @Override
    public void onDismissed() {
        if (mIsDismissed) {
            return;
        }
        mIsDismissed = true;
        mSelectedCardId = null;
        mHandler.removeCallbacks(mSelectionRunnable);
        mWalletClient.notifyWalletDismissed();
        mWalletClient.removeWalletServiceEventListener(this);
        mWalletView.animateDismissal();
    }

    /**
     * Implements {@link GlobalActionsPanelPlugin.PanelViewController}. Invoked when the device is
     * either locked or unlocked while the wallet is visible.
     */
    @Override
    public void onDeviceLockStateChanged(boolean deviceLocked) {
        if (mIsDismissed || mIsDeviceLocked == deviceLocked) {
            // Disregard repeat events
            return;
        }
        mIsDeviceLocked = deviceLocked;
        // Cards are re-queried because the wallet application may wish to change card art, icons,
        // text, or other attributes depending on the lock state of the device.
        queryWalletCards();
    }

    /**
     * Query wallet cards from the client and display them on screen.
     */
    void queryWalletCards() {
        if (mIsDismissed) {
            return;
        }
        if (!mHasRegisteredListener) {
            // Listener is registered even when device is locked. Should only be registered once.
            mWalletClient.addWalletServiceEventListener(this);
            mHasRegisteredListener = true;
        }
        if (mIsDeviceLocked && !mWalletClient.isWalletFeatureAvailableWhenDeviceLocked()) {
            mWalletView.showDeviceLockedMessage();
            return;
        }
        mWalletView.hideErrorMessage();
        int cardWidthPx = mWalletCardCarousel.getCardWidthPx();
        int cardHeightPx = mWalletCardCarousel.getCardHeightPx();
        int iconSizePx = mWalletView.getIconSizePx();
        GetWalletCardsRequest request =
                new GetWalletCardsRequest(cardWidthPx, cardHeightPx, iconSizePx, MAX_CARDS);
        mWalletClient.getWalletCards(mExecutor, request, this);
    }

    /**
     * Implements {@link QuickAccessWalletClient.OnWalletCardsRetrievedCallback}. Called when cards
     * are retrieved successfully from the service. This is called on {@link #mExecutor}.
     */
    @Override
    public void onWalletCardsRetrieved(GetWalletCardsResponse response) {
        List<WalletCard> walletCards = response.getWalletCards();
        // TODO: if walletCards is empty, show empty state view
        List<WalletCardViewInfo> data = new ArrayList<>(walletCards.size());
        for (WalletCard card : walletCards) {
            data.add(new QAWalletCardViewInfo(card));
        }
        // Get on main thread for UI updates
        mWalletView.post(() -> {
            if (mIsDismissed) {
                return;
            }
            boolean animate = mWalletCardCarousel.setData(data, response.getSelectedIndex());
            mWalletView.showCardCarousel(animate);
        });
    }

    /**
     * Implements {@link QuickAccessWalletClient.OnWalletCardsRetrievedCallback}. Called when there
     * is an error during card retrieval. This will be run on the {@link #mExecutor}.
     */
    @Override
    public void onWalletCardRetrievalError(GetWalletCardsError error) {
        mWalletView.post(() -> {
            if (mIsDismissed) {
                return;
            }
            mWalletView.showErrorMessage(error.getMessage());
        });
    }

    /**
     * Implements {@link QuickAccessWalletClient.WalletServiceEventListener}. Called when the wallet
     * application propagates an event, such as an NFC tap, to the quick access wallet view.
     */
    @Override
    public void onWalletServiceEvent(WalletServiceEvent event) {
        if (mIsDismissed) {
            return;
        }
        switch (event.getEventType()) {
            case WalletServiceEvent.TYPE_NFC_PAYMENT_STARTED:
                mPluginCallbacks.dismissGlobalActionsMenu();
                onDismissed();
                break;
            case WalletServiceEvent.TYPE_WALLET_CARDS_UPDATED:
                queryWalletCards();
                break;
            default:
                Log.w(TAG, "onWalletServiceEvent: Unknown event type");
        }
    }

    /**
     * Implements {@link WalletCardCarousel.OnSelectionListener}. Called when the user selects a
     * card from the carousel by scrolling to it.
     */
    @Override
    public void onCardSelected(WalletCardViewInfo card) {
        if (mIsDismissed) {
            return;
        }
        mSelectedCardId = card.getCardId();
        selectCard();
    }

    /**
     * Implements {@link WalletCardCarousel.OnSelectionListener}. Called when the user taps on the
     * view outside of a card target which should cause the wallet to be dismissed.
     */
    @Override
    public void onDismissGesture() {
        if (mIsDismissed) {
            return;
        }
        mPluginCallbacks.dismissGlobalActionsMenu();
        onDismissed();
    }

    private void selectCard() {
        mHandler.removeCallbacks(mSelectionRunnable);
        String selectedCardId = mSelectedCardId;
        if (mIsDismissed || selectedCardId == null) {
            return;
        }
        mWalletClient.selectWalletCard(new SelectWalletCardRequest(selectedCardId));
        // Re-selecting the card keeps the connection bound so we continue to get service events
        // even if the user keeps it open for a long time.
        mHandler.postDelayed(mSelectionRunnable, SELECTION_DELAY_MILLIS);
    }

    /**
     * Implements {@link WalletCardCarousel.OnSelectionListener}. Called when the user clicks on a
     * card.
     */
    @Override
    public void onCardClicked(WalletCardViewInfo card) {
        if (mIsDismissed) {
            return;
        }
        PendingIntent pendingIntent = ((QAWalletCardViewInfo) card).mWalletCard.getPendingIntent();
        mPluginCallbacks.startPendingIntentDismissingKeyguard(pendingIntent);
        mPluginCallbacks.dismissGlobalActionsMenu();
        onDismissed();
    }

    private class QAWalletCardViewInfo implements WalletCardViewInfo {

        private final WalletCard mWalletCard;
        private final Drawable mCardDrawable;
        private final Drawable mIconDrawable;

        /**
         * Constructor is called on background executor, so it is safe to load drawables
         * synchronously.
         */
        QAWalletCardViewInfo(WalletCard walletCard) {
            mWalletCard = walletCard;
            mCardDrawable = mWalletCard.getCardImage().loadDrawable(mPluginContext);
            Icon icon = mWalletCard.getCardIcon();
            mIconDrawable = icon == null ? null : icon.loadDrawable(mPluginContext);
        }

        @Override
        public String getCardId() {
            return mWalletCard.getCardId();
        }

        @Override
        public Drawable getCardDrawable() {
            return mCardDrawable;
        }

        @Override
        public CharSequence getContentDescription() {
            return mWalletCard.getContentDescription();
        }

        @Override
        public Drawable getIcon() {
            return mIconDrawable;
        }

        @Override
        public CharSequence getText() {
            return mWalletCard.getCardLabel();
        }
    }
}

File: lib/WalletPanelViewController.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.plugin.globalactions.wallet

import android.animation.TimeInterpolator
import android.app.PendingIntent
import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.LinearInterpolator
import android.widget.TextView
import com.android.systemui.plugin.globalactions.wallet.common.CardDimens
import com.android.systemui.plugin.globalactions.wallet.reactive.CachingEventSource
import com.android.systemui.plugin.globalactions.wallet.reactive.EventStream
import com.android.systemui.plugin.globalactions.wallet.reactive.Subscription
import com.android.systemui.plugin.globalactions.wallet.reactive.events
import com.android.systemui.plugin.globalactions.wallet.reactive.mergeEvents
import com.android.systemui.plugin.globalactions.wallet.reactive.setCancelAction
import com.android.systemui.plugin.globalactions.wallet.view.R
import com.android.systemui.plugin.globalactions.wallet.view.WalletBackgroundDrawable
import com.android.systemui.plugin.globalactions.wallet.view.common.TopLevelViewModel
import com.android.systemui.plugin.globalactions.wallet.view.walletview.ICON_SIZE_DP
import com.android.systemui.plugin.globalactions.wallet.view.walletview.TopLevelWalletView
import com.android.systemui.plugins.GlobalActionsPanelPlugin
import kotlin.math.roundToInt

private const val TAG = "WalletPanelViewCtrlr"

private const val SCROLL_X_ANIMATION_OFFSET_PX = 120f
private const val ANIMATION_DURATION = 300L

private val outInterpolator: TimeInterpolator = AccelerateInterpolator()

/** View controller for the Quick Access Wallet. */
class WalletPanelViewController(
        context: Context,
        callbacks: GlobalActionsPanelPlugin.Callbacks,
        deviceLocked: Boolean,
        panelComponentFactory: WalletPanelComponentFactory
) : GlobalActionsPanelPlugin.PanelViewController {

    private val panelView: View =
        LayoutInflater.from(context).inflate(R.layout.global_actions_activity, null)
    private val panelClicks: EventStream<Unit> = events {
        setCancelAction { panelView.setOnClickListener(null) }
        panelView.setOnClickListener { emitEvent(Unit) }
    }
    private val walletView: TopLevelWalletView<Drawable> =
            panelView.requireViewById(R.id.view_pager)
    private val errorView: TextView = panelView.requireViewById(R.id.error_view)
    private val backgroundDrawable = WalletBackgroundDrawable(context)
    private val lockEvents = CachingEventSource(deviceLocked)
    private val panelComponent: WalletUiComponent = run {
        val pendingIntentSender = object : PendingIntentSender {
            override fun sendPendingIntent(pendingIntent: PendingIntent): Boolean {
                callbacks.startPendingIntentDismissingKeyguard(pendingIntent)
                return true
            }
        }
        val cardDimens = CardDimens(
                walletView.cardWidthPx,
                walletView.cardHeightPx,
                (ICON_SIZE_DP * context.resources.displayMetrics.density).roundToInt()
        )
        val panelCallbacks = object : PanelCallbacks {
            override fun dismissGlobalActionsMenu() {
                callbacks.dismissGlobalActionsMenu()
            }

            override fun showErrorMessage(error: CharSequence?) {
                showError(error ?: context.resources.getString(R.string.error_generic))
            }

            override fun populateUi(cards: List<TopLevelViewModel<Drawable>>, selectedIndex: Int) {
                walletView.visibility = View.VISIBLE
                walletView.setCards(cards, selectedIndex)
                backgroundDrawable.darken()
                errorView.visibility = View.GONE
            }
        }
        val dismissRequests = mergeEvents(panelClicks, walletView.dismissEvents)

        panelComponentFactory.createPanelComponent(
                callbacks = panelCallbacks,
                cardDimens = cardDimens,
                pendingIntentSender = pendingIntentSender,
                uiDismissRequests = dismissRequests,
                lockEvents = lockEvents,
                uiCardSelections = walletView.selectEvents,
                uiCardClicks = walletView.clickEvents,
                uiSettingsButtonClicks = walletView.settingsEvents
        )
    }

    private val uiSubscription: Subscription

    init {
        // Connect UI to backend, and kick-off backend queries
        uiSubscription = panelComponent.uiController.connect()
    }

    override fun onDismissed() {
        uiSubscription.cancel()
        if (walletView.visibility == View.VISIBLE) {
            for (i in 0 until walletView.childCount) {
                val view = walletView.getChildAt(i)
                view.animate().apply {
                    duration = ANIMATION_DURATION
                    interpolator = outInterpolator
                    alpha(0f)
                    translationX(SCROLL_X_ANIMATION_OFFSET_PX)
                    start()
                }
            }
        }
        if (errorView.visibility == View.VISIBLE) {
            errorView.animate().apply {
                alpha(0f)
                duration = ANIMATION_DURATION
                interpolator = outInterpolator
                start()
            }
        }
    }

    override fun getPanelContent(): View = panelView

    override fun onDeviceLockStateChanged(locked: Boolean) = lockEvents.emitEvent(locked)

    override fun getBackgroundDrawable(): Drawable = backgroundDrawable

    private fun showError(errorText: CharSequence) {
        errorView.text = errorText
        if (errorView.visibility != View.VISIBLE) {
            errorView.visibility = View.VISIBLE
            errorView.alpha = 0f
            errorView.animate().apply {
                alpha(1f)
                duration = ANIMATION_DURATION
                interpolator = LinearInterpolator()
                start()
            }
        }
        walletView.visibility = View.INVISIBLE
    }
}
